### DISCUSSION TITLE
How to run multiple GATT servers

### DISCUSSION THREAD

**Discussion by evan361425** [google/bumble#572] [Q&A] [answered]
I have tried sharing `Transport` like:

```python
async with await open_transport_or_link(link_name) as transport:
  start_device(create_device1(transport))
  start_device(create_device2(transport))
```

Both `create_device` create device by calling `Device.from_config_with_hci(config, transport.source, transport.sink)`
and `start_device` await both `device.power_on()` and `device.start_advertising()`.

But got error

```text
File "./site-packages/bumble/host.py", line 246, in reset
    response.return_parameters.supported_commands, 'little'
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'int' object has no attribute 'supported_commands'
```

I also tried create `Transport` in each device, but still got error:

```text
File "./site-packages/watchdog/observers/fsevents.py", line 307, in run
    _fsevents.add_watch(self, self.watch, self.events_callback, self.pathnames)
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Cannot add watch <ObservedWatch: path='printer_status', is_recursive=True> - it is already scheduled
```

It turns out that only first device are listening.

My question is what is the correct way to start multiple servers in single process?

**Comment by zxzxwu** [comment, marked as answer]
Transport cannot be shared by multiple hosts. If you have multiple transports, you can run independent host on each transport like https://github.com/google/bumble/blob/main/examples/run_csis_servers.py.

In fact, running multiple GATT servers on a single device is impossible, because all of them share the same ATT L2CAP channel. So, I think you need to check what's your actual use case and solve them in the other way, instead of reusing the transport.

**Comment by barbibulle** [comment]
As @zxzxwu points out, you can only have one Device/Host per transport.
But multiple GATT servers are possible, in some way. There are two ways to do that.
The first option, if what you need is to have a GATT server that exposes the same services and characteristics to all clients, but responds with different values for different clients, is not really to have multiple GATT servers, but rather use the default GATT server (the one automatically created for you when you create the Device object), but have your GATT handlers return results based on which client is doing a read or write: characteristic delegates are methods that take a `connection` as their first parameter. Each client will have a different connection, so your handler can return different things to different clients.
The second option, more advanced (not recommended, as it isn't really a very common use case), is to really have one GATT server per connection. You can do that, because each connection has a `gatt_server` property. By default, when a connection is created, the connection's `gatt_server` property is set to the Device's shared `gatt_server` property. But you're free to create a different `gatt_server.Server` instance and replace the connection's `gatt_server` property with that. If you do that, it's then really up to you to manage the lifetime of that gatt server.

**Comment by evan361425** [comment]
@zxzxwu  @barbibulle  Thanks a lot for your reply.

I should first state out my use case. I serve two devices by running this python code twice:

```python
import asyncio
from sys import argv

from bumble.core import AdvertisingData
from bumble.device import Connection, Device, DeviceConfiguration
from bumble.hci import Address
from bumble.transport import open_transport_or_link
from bumble.transport.common import Transport


class Listener(Device.Listener, Connection.Listener):
    def __init__(self, device):
        self.device = device

    def on_connection(self, connection):
        print(f"=== Connected to {connection}")
        connection.listener = self

    def on_disconnection(self, reason):
        print(f"### Disconnected, reason={reason}")


def create_device(name: str, address: str, transport: Transport) -> Device:
    config = DeviceConfiguration(
        name=name,
        address=Address(address),
        advertising_data=bytes(
            AdvertisingData(
                [(AdvertisingData.COMPLETE_LOCAL_NAME, bytes(name, "utf-8"))]
            )
        ),
    )
    device = Device.from_config_with_hci(config, transport.source, transport.sink)
    device.listener = Listener(device)

    return device


async def main():
    transport = await open_transport_or_link("android-netsim")

    device = create_device(argv[1], argv[2], transport)
    await device.power_on()
    await device.start_advertising(auto_restart=True)
    print(f"start device {argv[1]}")

    await transport.source.terminated


asyncio.run(main())
```

for example:

```bash
$ python run_gatt.py Device1 F0:F0:F0:F0:F0:F0
start device Device1
# Another terminal
$ python run_gatt.py Device2 F0:F0:F0:F0:F0:F1
start device Device2
```

After execute the code twice, I can have two device listen on `android-netsim` (my app can discovery these two).

Now, I want to use single python process to start up multiple server:

```python
async def main():
    transports = []

    for name, address in zip(argv[1::2], argv[2::2]):
        transport = await open_transport_or_link("android-netsim")
        transports.append(transport)

        device = await create_device(name, address)
        await device.power_on()
        await device.start_advertising(auto_restart=True)
        print(f'start device {argv[1]}')

    await asyncio.gather(*[e.source.terminated for e in transports])
```

I create transport for each device from your advices

> you can run independent host on each transport @zxzxwu
>
> you can only have one Device/Host per transport @barbibulle 

But I got the error and my app can only discovery first device (`Device1`):

```bash
$ python run_gatt.py Device1 F0:F0:F0:F0:F0:F0 Device2 F0:F0:F0:F0:F0:F1
start device Device1
ERROR:fsevents:Unhandled exception in FSEventsEmitter
Traceback (most recent call last):
  File "./site-packages/watchdog/observers/fsevents.py", line 307, in run
    _fsevents.add_watch(self, self.watch, self.events_callback, self.pathnames)
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
RuntimeError: Cannot add watch <ObservedWatch: path='printer_status', is_recursive=True> - it is already scheduled
start device Device2
```

**Reply by zxzxwu** [reply]
I cannot reproduce your issue. Are you using scripts with other modules? Your log looks very different.

The code looks wrong as well. `device = await create_device(name, address)` shouldn't work because `create_device` is not async.

**Reply by barbibulle** [reply]
It looks like your error has something to do with how you use FSEvents watchers, not something in the Bumble library.

**Reply by evan361425** [reply]
Thanks @zxzxwu, I've fixed the code, and I really appreciate your correction!  
@barbibulle, you were spot on—it turns out my code was watching the file multiple times. Took me a while to figure it out.  
Once again, thanks to both of you for your previous answers on GATT and Transport. They helped me a lot!
